pkg/async/schedule/flytescheduler/workflow_executor.go

```diff
@@ -0,0 +1,234 @@
+				logger.Errorf(ctx, "unable to close the file %v after writing due to %v", fileName, err)
+			}
+		}(f)
+		// Just log the error.
+		if err != nil {
+			logger.Errorf(ctx, "unable to create a writable snapshot file %v due to %v", fileName, err)
+		}
+		err = w.snapShotReaderWriter.WriteSnapshot(f, w.snapshot)
+		// Just log the error
+		if err != nil {
+			logger.Errorf(ctx, "unable to write the snapshot to file %v due to %v", fileName, err)
+		}
+		time.Sleep(snapshotWriterSleepTime * time.Second)
+	}
+}
+
+func (w *workflowExecutor) CatchUpAllSchedules(ctx context.Context, schedules []models.SchedulableEntity, toTime time.Time) error {
+	for _, s := range schedules {
+		err := w.CatchUpSingleSchedule(ctx, s, toTime)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+
+func (w *workflowExecutor) CatchUpSingleSchedule(ctx context.Context, s models.SchedulableEntity, toTime time.Time) error {
+	nameOfSchedule := w.goGfInterface.GetScheduleName(s)
+	lastExecTime := w.snapshot.GetLastExecutionTime(nameOfSchedule)
+	// Get the jitter value
+	workflowExecConfig := w.config.ApplicationConfiguration().GetSchedulerConfig().WorkflowExecutorConfig.FlyteWorkflowExecutorConfig
+	jitter := time.Duration(workflowExecConfig.JitterValue) * time.Second
+	schedulerEpochTime := workflowExecConfig.SchedulerEpochTime
+
+	var fromTime time.Time
+	// If the scheduler epoch time ise set then use that to catch up till the current time
+	// else check the last exec time for the schedule and if that also is not set then use current time with
+	// configured jitter
+	if !schedulerEpochTime.IsZero() {
+		fromTime = schedulerEpochTime
+	} else {
+		if !lastExecTime.IsZero() {
+			fromTime = lastExecTime
+		} else {
+			fromTime = time.Now().Add(-jitter)
+		}
+	}
+
+	var catchUpTimes []time.Time
+	var err error
+	catchUpTimes, err = w.goGfInterface.GetCatchUpTimes(s, fromTime, toTime)
+	if err != nil {
+		return err
+	}
+	var catchupTime time.Time
+	for _, catchupTime = range catchUpTimes {
+		err := fire(ctx, w.executionManager, catchupTime, s)
+		if err != nil {
+			logger.Errorf(ctx, "unable to fire the schedule %v at %v time due to %v", s, catchupTime, err)
+		} else {
+			w.snapshot.UpdateLastExecutionTime(nameOfSchedule, catchupTime)
+		}
+	}
+
+	return nil
+}
+
+func fire(ctx context.Context, executionManager mgInterfaces.ExecutionInterface, scheduledTime time.Time,
+	s models.SchedulableEntity) error {
+
+	literalsInputMap := map[string]*core.Literal{}
+	literalsInputMap[s.KickoffTimeInputArg] = &core.Literal{
+		Value: &core.Literal_Scalar{
+			Scalar: &core.Scalar{
+				Value: &core.Scalar_Primitive{
+					Primitive: &core.Primitive{
+						Value: &core.Primitive_Datetime{
+							Datetime: timestamppb.New(scheduledTime),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	executionRequest := admin.ExecutionCreateRequest{
+		Project: s.Project,
+		Domain:  s.Domain,
+		Name:    "f" + strings.ReplaceAll(uuid.New().String(), "-", "")[:19],
```

Comment:
don't we want this name to be deterministic based on the kickoff/scheduled time?

Reply:
I was trying to keep this consistent with other places where we are generating the UUID for the executions .
The UUID on the details does give us the scheduleTime in its metadata field .

Do we want something 

like instead of this  `f09f1395a2d824cf9800` to have  this `2021-08-04T17:07:37+05:30-f09f1395a2d824cf9800`
or some other form.

Reply:
a hash of the input time is fine. if we generate a new uuid every time though we don't get idempotency from repeated calls to trigger a scheduled execution

Reply:
we should not use UUID, to make it idempotent, it has to be deterministic